cpp/src/arrow/util/future.cc

```diff
@@ -89,7 +90,7 @@ class ConcreteFutureImpl : public FutureImpl {
       case ShouldSchedule::IfUnfinished:
         return !in_add_callback;
       case ShouldSchedule::IfDifferentExecutor:
-        return !callback_record.options.executor->OwnsThisThread();
+        return !(callback_record.options.executor->IsCurrentExecutor());
```

Comment:
Why do we need this change?
Can we just disable (or error) this feature without threading?

Reply:
I'm not sure - I didn't really know what this feature was used for and who uses it, so I made sure I didn't break it.